packages/gatsby/src/redux/actions/restricted.js

```diff
@@ -0,0 +1,276 @@
+ *     type AuthorJson implements Node @dontInfer {
+ *       name: String!
+ *       birthday: Date! @dateformat(locale: "ru")
+ *     }
+ *   `
+ *   createTypes(typeDefs)
+ * }
+ *
+ * // using Gatsby Type Builder API
+ * exports.sourceNodes = ({ actions, schema }) => {
+ *   const { createTypes } = actions
+ *   const typeDefs = [
+ *     schema.buildObjectType({
+ *       name: 'MarkdownRemark',
+ *       fields: {
+ *         frontmatter: 'Frontmatter!'
+ *       },
+ *       interfaces: ['Node'],
+ *       extensions: {
+ *         infer: true,
+ *       },
+ *     }),
+ *     schema.buildObjectType({
+ *       name: 'Frontmatter',
+ *       fields: {
+ *         title: {
+ *           type: 'String!',
+ *           resolve(parent) {
+ *             return parent.title || '(Untitled)'
+ *           }
+ *         },
+ *         author: {
+ *           type: 'AuthorJson'
+ *           extensions: {
+ *             link: {},
+ *           },
+ *         }
+ *         date: {
+ *           type: 'Date!'
+ *           extensions: {
+ *             dateformat: {},
+ *           },
+ *         },
+ *         published: 'Boolean!',
+ *         tags: '[String!]!',
+ *       }
+ *     }),
+ *     schema.buildObjectType({
+ *       name: 'AuthorJson',
+ *       fields: {
+ *         name: 'String!'
+ *         birthday: {
+ *           type: 'Date!'
+ *           extensions: {
+ *             dateformat: {
+ *               locale: 'ru',
+ *             },
+ *           },
+ *         },
+ *       },
+ *       interfaces: ['Node'],
+ *       extensions: {
+ *         infer: false,
+ *       },
+ *     }),
+ *   ]
+ *   createTypes(typeDefs)
+ * }
+ */
+actions.createTypes = (
+  types:
+    | string
+    | GraphQLOutputType
+    | GatsbyGraphQLType
+    | Array<string | GraphQLOutputType | GatsbyGraphQLType>,
+  plugin: Plugin,
+  traceId?: string
+) => {
+  return {
+    type: `CREATE_TYPES`,
+    plugin,
+    traceId,
+    payload: types,
+  }
+}
+
+const withDeprecationWarning = ({
+  api,
+  actions,
+  actionNames,
+  alternativeAPI,
+}) =>
+  actionNames.reduce((acc, action) => {
+    acc[action] = (...args) => {
+      report.warn(
+        `Calling \`${action}\` in the \`${api}\` API is deprecated. ` +
+          `Please use: ${alternativeAPI.map(a => `\`${a}\``).join(`, `)}.`
+      )
+      return actions[action](...args)
+    }
+    return acc
+  }, {})
+
+const withErrorMessage = ({ api, actionNames, alternativeAPI }) =>
+  actionNames.reduce((acc, action) => {
+    acc[action] = () => {
+      report.error(
+        `\`${action}\` is not available in the \`${api}\` API. ` +
+          `Please use: ${alternativeAPI.map(a => `\`${a}\``).join(`, `)}.`
+      )
+      // FIXME: Dummy error action - required because action creators
+      // must return an action. Better would be to have a logging middleware
+      // so we could just `return log({ message, type: 'error' })`
+      return { type: `DISPATCH_ERROR` }
```

Comment:
Maybe we could return empty array or adjust our redux middleware to skip `null` or `undefined` "actions"?

Reply:
yes i think easiest is probably to use middleware we already have and just return a thunk that does not dispatch.